include/lbann/execution_contexts/execution_context.hpp

```diff
@@ -0,0 +1,143 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2016, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_EXECUTION_CONTEXT_HPP
+#define LBANN_EXECUTION_CONTEXT_HPP
+
+#include "lbann/base.hpp"
+#include "lbann/comm.hpp"
+#include "lbann/io/persist.hpp"
+#include "lbann/utils/threads/thread_pool.hpp"
+
+namespace lbann {
+
+// Forward-declare this.
+class trainer;
+
+struct termination_criteria {
```

Comment:
It's almost more intuitive to have stopping criteria posed in a polymorphic manner that mirrors `execution_context` and `training_algorithm`. For stochastic algorithms, one might have minibatch or epoch limits. For batched algorithms, you might have iteration limits. Or it might be an accuracy target or some other such thing. So I agree that it should be tied to the training algorithm, but it should not be internal. Users may need to set, inspect, or manipulate the stopping criteria. The other option would be the Trilinos/ROL-y option of having a "catch-all" criteria class that lists every possible thing that one might use with only a subset being tested by particular algorithms (I do not like this approach and find it somewhat misleading (e.g., you specify a residual tolerance because it's listed in the stopping criteria class but the algorithm you chose doesn't test that residual or something)).

Reply:
I see, so part of implementing a new `training_algorithm` would be writing the custom termination criteria that it could use. There would be a few basic termination criteria (e.g. based on step count, epoch count, or runtime) that would be used by many algorithms, but maybe not all. This introduces the issue that you need to know which termination criteria are compatible with which algorithms. This could be workable, we're just taking complexity out of the algorithm implementation and making the user deal with it.

Reply:
Sure. Ideally each `training_algorithm` would know which stopping criteria it supports and can give you a prototypical copy of this structure. Since it will be filled from a prototext message anyway (`google::protobuf::Message`, none of this `lbann_data::LbannPB` nonsense), the polymorphic path basically writes itself. :)